core/modules/util/EventThread.h

```diff
@@ -0,0 +1,195 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2015 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ *  @author: John Gates,
+ */
+
+#ifndef LSST_QSERV_UTIL_EVENTTHREAD_H_
+#define LSST_QSERV_UTIL_EVENTTHREAD_H_
+
+// System headers
+#include <atomic>
+#include <cassert>
+#include <deque>
+#include <thread>
+#include <vector>
+
+// Qserv headers
+#include "util/Command.h"
+
+namespace lsst {
+namespace qserv {
+namespace util {
+
+/// A queue of Commands meant to drive an EventThread.
+///
+class CommandQueue {
+public:
+    using Ptr = std::shared_ptr<CommandQueue>;
+    virtual ~CommandQueue() {};
+    /// Queue a command object in a thread safe way and signal any threads
+    /// waiting on the queue that a command is available.
+    virtual void queCmd(Command::Ptr const& cmd) {
+        std::lock_guard<std::mutex> lock(_mx);
+        _qu.push_back(cmd);
+        notify(_qu.size() > 1);
+    };
+
+    /// Get a command off the queue.
+    /// If wait is true, wait until a message is available.
+    virtual Command::Ptr getCmd(bool wait=true) {
+        std::unique_lock<std::mutex> lock(_mx);
+        if (wait) {
+            _cv.wait(lock, [this](){return !_qu.empty();});
+        }
+        if (_qu.empty()) {
+            return nullptr;
+        }
+        auto cmd = _qu.front();
+        _qu.pop_front();
+        return cmd;
+    };
+
+    /// Notify all threads waiting on this queue, or just 1 if all is false.
+    virtual void notify(bool all=true) {
+        if (all) {
+            _cv.notify_all();
+        } else {
+            _cv.notify_one();
+        }
+    }
+
+    virtual void commandStart(Command::Ptr const&) {};
+    virtual void commandFinish(Command::Ptr const&) {};
+protected:
+    std::deque<Command::Ptr> _qu{};
+    std::condition_variable  _cv{};
+    std::mutex               _mx{};
+};
+
+/// An event driven thread.
+/// Thread must be started with run(). Stop the thread by calling queEnd().
+class EventThread {
+public:
+    enum { HALT = -1000 };
```

Comment:
The ThreadPool worked out very smoothly until it came time to shut a thread down. Having Command::action() return a specific int is a solution, but I don't like it very much. I've got another idea, that might be a little better. If I made it a class, it would need to be defined before Util::Command, and I really don't like that.